src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -666,131 +809,88 @@ func (s *commitLogSource) newReadCommitlogPredAndMostRecentSnapshotByBlockShard(
+	worker *accumulateWorker,
 ) {
-	nsCtx := namespace.NewContextFrom(ns)
-	for arg := range ec {
-		var (
-			series     = arg.series
-			dp         = arg.dp
-			unit       = arg.unit
-			annotation = arg.annotation
-			blockStart = arg.blockStart
-		)
-
+	ctx := context.NewContext()
+	for input := range worker.inputCh {
 		var (
-			unmergedShard      = unmerged[series.Shard].series
-			unmergedSeries, ok = unmergedShard.Get(series.ID)
+			namespace  = input.namespace
+			entry      = input.series
+			shard      = input.shard
+			dp         = input.dp
+			unit       = input.unit
+			annotation = input.annotation
 		)
-		if !ok {
-			unmergedSeries = metadataAndEncodersByTime{
-				id:       series.ID,
-				tags:     series.Tags,
-				encoders: make(map[xtime.UnixNano][]encoder)}
-			// Have to use unsafe because we don't want to copy the IDs we put
-			// into this map because its lifecycle is much shorter than that of
-			// the IDs we're putting into it so copying would waste too much
-			// memory unnecessarily, and we don't want to finalize the IDs for the
-			// same reason.
-			unmergedShard.SetUnsafe(
-				series.ID, unmergedSeries,
-				SetUnsafeOptions{NoCopyKey: true, NoFinalizeKey: true})
+		if !s.shouldAccmulateForTime(namespace, shard, dp.Timestamp) {
+			worker.datapointsSkipped++
+			continue
 		}
 
-		var (
-			err            error
-			blockStartNano = xtime.ToUnixNano(blockStart)
-			unmergedBlock  = unmergedSeries.encoders[blockStartNano]
-			wroteExisting  = false
-		)
-		for i := range unmergedBlock {
-			// TODO(r): Write unit test to ensure that different values that arrive
-			// later in the commit log will upsert the previous value when bootstrapping
-			// Tracking with issue: https://github.com/m3db/m3/issues/898
-			if unmergedBlock[i].lastWriteAt.Before(dp.Timestamp) {
-				unmergedBlock[i].lastWriteAt = dp.Timestamp
-				err = unmergedBlock[i].enc.Encode(dp, unit, annotation)
-				wroteExisting = true
-				break
-			}
-		}
-		if !wroteExisting {
-			enc := encoderPool.Get()
-			enc.Reset(blockStart, blopts.DatabaseBlockAllocSize(), nsCtx.Schema)
-
-			err = enc.Encode(dp, unit, annotation)
-			if err == nil {
-				unmergedBlock = append(unmergedBlock, encoder{
-					lastWriteAt: dp.Timestamp,
-					enc:         enc,
-				})
-				unmergedSeries.encoders[blockStartNano] = unmergedBlock
-			}
-		}
+		worker.datapointsRead++
+
+		_, err := entry.Series.Write(ctx, dp.Timestamp, dp.Value,
+			unit, annotation, series.WriteOptions{
+				SchemaDesc: namespace.namespaceContext.Schema,
+				// NB(r): Make sure this is the series we originally
+				// checked out for writing too (which should be guaranteed
+				// by the fact during shard tick we do not expire any
+				// series unless they are bootstrapped).
+				MatchUniqueIndex:      true,
+				MatchUniqueIndexValue: entry.UniqueIndex,
+			})
 		if err != nil {
-			workerErrs[workerNum]++
+			// NB(r): Debug log since this could be very noisy if it
```

Comment:
Would it be worth sampling this and running on info?

Reply:
Maybe I'll do this once per worker?  That could help.